config/config.go

```diff
@@ -0,0 +1,163 @@
+	if os.IsNotExist(err) {
+		file, err := os.Create(filePath)
+		if err != nil {
+			return nil, err
+		}
+		defer file.Close()
+		return generateConfig(filePath)
+	} else if err != nil {
+		return nil, err
+	}
+	return validateConfig(filePath)
+}
+
+func generateConfig(configPath string) (*ViewerConfig, error) {
+	viewerConfig := &proto.ViewerConfig{}
+
+	viewerConfig.Url = &DefaultURL
+	viewerConfig.CssSelector = &proto.CSSSelector{
+		Selector: &DefaultCSSSelector.Selector,
+		Position: &DefaultCSSSelector.Position,
+	}
+	viewerConfig.BrowserWindowCount = &DefaultBrowserCount
+	viewerConfig.UserDataDirPrefix = &DefaultUserDataDirPrefix
+
+	DefaultInputWindowPosition := proto.ViewerConfig_DEFAULT_BOTTOM
+	viewerConfig.InputWindowPosition = &DefaultInputWindowPosition
+
+	rootWindowLayout := &proto.Layout{
+		Width:  &DefaultRootWindowWidth,
+		Height: &DefaultRootWindowHeight,
+	}
+	viewerConfig.RootWindowConfig = &proto.RootWindowConfig{
+		Layout: rootWindowLayout,
+	}
+
+	marshalOpts := prototext.MarshalOptions{Multiline: true, Indent: "\t"}
+	out, err := marshalOpts.Marshal(viewerConfig)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to encode aso_sxs_viewer config %s", err)
+	}
+	out = append(out, []byte(BrowserWindowExample)...)
+	if err := ioutil.WriteFile(configPath, out, 0644); err != nil {
+		return nil, fmt.Errorf("Failed to write aso_sxs_viewer config %s", err)
+	}
+	return &ViewerConfig{*viewerConfig}, nil
+}
+
+// validateConfig validate the given configFile. Any assumptions made to handle essential empty fields are written back to the config file.
+func validateConfig(configPath string) (*ViewerConfig, error) {
+	var writeChanges bool
+	in, err := ioutil.ReadFile(configPath)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to read aso_sxs_viewer config %s", err)
+	}
+
+	viewerConfig := &proto.ViewerConfig{}
+
+	if err := prototext.Unmarshal(in, viewerConfig); err != nil {
+		return nil, fmt.Errorf("Failed to parse aso_sxs_viewer config %s", err)
+	}
+
+	if viewerConfig == nil {
+		return nil, fmt.Errorf("Found an empty config file, please delete the existing config and try again")
+	}
+
+	if viewerConfig.BrowserWindowCount == nil {
+		viewerConfig.BrowserWindowCount = &DefaultBrowserCount
+		writeChanges = true
+	} else if viewerConfig.GetBrowserWindowCount() <= 0 {
+		return nil, fmt.Errorf("A positive browser_window_count field is required in the config.textproto file")
+	}
+
+	// If both the CSSSelector and url are nil, try to populate them using the first WindowOverrides
+	// Otherwise use the populate them using the default values.
+	// If only one of the fields is non nil throw an error.
+	if CSSSelector, url := viewerConfig.GetCssSelector().GetSelector(), viewerConfig.GetUrl(); CSSSelector == "" && url == "" {
+		if windowsOverrides := viewerConfig.GetWindowOverrides(); windowsOverrides != nil && windowsOverrides[0].GetCssSelector().GetSelector() != "" && windowsOverrides[0].GetUrl() != "" {
```

Comment:
wouldn't this crash if no window override is present?

I also don't really like the idea of using the first window_override for everything.

For each n windows, try to read the window_override, if url/css is not mentioned, check the default. If neither is mentioned fail.

That way, if n = 2, either user needs to specify defaults or user needs to specify BOTH window_overrides.